destination_middleware.go

```diff
@@ -236,3 +254,208 @@ func (d *destinationWithRateLimit) Write(ctx context.Context, recs []Record) (in
+	}
+	if len(d.RecordEncoders) == 0 {
+		d.RecordEncoders = d.DefaultEncoders()
+	}
+
+	// sort converters and encoders by name to ensure we can binary search them
+	sort.Slice(d.RecordConverters, func(i, j int) bool { return d.RecordConverters[i].Name() < d.RecordConverters[j].Name() })
+	sort.Slice(d.RecordEncoders, func(i, j int) bool { return d.RecordEncoders[i].Name() < d.RecordEncoders[j].Name() })
+
+	return &destinationWithRecordFormat{
+		Destination: impl,
+		defaults:    d,
+	}
+}
+
+type destinationWithRecordFormat struct {
+	Destination
+	defaults DestinationWithRecordFormat
+
+	formatter recordFormatter
+}
+
+func (d *destinationWithRecordFormat) converterNames() []string {
+	names := make([]string, len(d.defaults.RecordConverters))
+	i := 0
+	for _, c := range d.defaults.RecordConverters {
+		names[i] = c.Name()
+		i++
+	}
+	return names
+}
+
+func (d *destinationWithRecordFormat) encoderNames() []string {
+	names := make([]string, len(d.defaults.RecordEncoders))
+	i := 0
+	for _, e := range d.defaults.RecordEncoders {
+		names[i] = e.Name()
+		i++
+	}
+	sort.Strings(names) // ensure we can binary search the slice
+	return names
+}
+
+func (d *destinationWithRecordFormat) formats() []string {
+	cs := d.converterNames()
+	es := d.encoderNames()
+
+	formats := make([]string, len(cs)*len(es))
+	for i, c := range d.converterNames() {
+		for j, e := range d.encoderNames() {
+			formats[(i*len(es))+j] = c + "/" + e
+		}
+	}
+	return formats
+}
+
+func (d *destinationWithRecordFormat) Parameters() map[string]Parameter {
+	return mergeParameters(d.Destination.Parameters(), map[string]Parameter{
+		configDestinationRecordFormat: {
+			Default:     d.defaults.DefaultRecordFormat,
+			Description: "TODO",
+			Validations: []Validation{
+				ValidationInclusion{List: d.formats()},
+			},
+		},
+		configDestinationRecordFormatOptions: {
+			Description: "TODO",
+		},
+	})
+}
+
+func (d *destinationWithRecordFormat) Configure(ctx context.Context, config map[string]string) error {
+	err := d.Destination.Configure(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	format := d.defaults.DefaultRecordFormat
+	if f, ok := config[configDestinationRecordFormat]; ok {
+		format = f
+	}
+
+	pair := strings.Split(format, "/")
+	if len(pair) != 2 {
+		return fmt.Errorf("invalid %s: expected %q to follow the structure \"type/subtype\"", configDestinationRecordFormat, format)
+	}
+
+	opt := d.parseFormatOptions(config[configDestinationRecordFormatOptions])
+	converter, err := d.parseConverter(pair[0], opt)
+	if err != nil {
+		return err
+	}
+	encoder, err := d.parseEncoder(pair[1], opt)
+	if err != nil {
+		return err
+	}
+
+	d.formatter = recordFormatter{
+		converter: converter,
+		encoder:   encoder,
+	}
+	return nil
+}
+
+func (d *destinationWithRecordFormat) parseFormatOptions(options string) map[string]string {
+	options = strings.TrimSpace(options)
+	if len(options) == 0 {
+		return nil
+	}
+
+	pairs := strings.Split(options, ",")
```

Comment:
we have splits that use `/`  `,` and `=` as separators, so it would be nice to have them as constants to know what they separate